[v258] musl specific patches

yuwata opened the pull request:
This contains backports from the main branch to allow building systemd v258 with musl.
If you want to build systemd v258 with musl, it is recommended that all commits in this branch are cherry-picked.

yuwata committed:
musl: glob-util: filter out . and .. even if GLOB_ALTDIRFUNC is not supported

musl neither support GLOB_ALTDIRFUNC nor GLOB_BRACE.
Let's make safe_glob() work even when GLOB_ALTDIRFUNC is not supported.
Currently, GLOB_BRACE is simply ignored when it is not supported.

(cherry picked from commit 07e437f569c68e218b2e7592dae458b0d50fbaad)

ChenQi1989 committed:
musl: add missing FTW_CONTINUE macro

This is to avoid build failures like below for musl.

  test-recurse-dir.c:23:24: error: ‘FTW_CONTINUE’ undeclared

Co-authored-by: Yu Watanabe <watanabe.yu+github@gmail.com>
(cherry picked from commit dd102894e7629b3fcbaf5ea3f26ed5f9d4c0cfbb)

esmil committed:
musl: add fallback parse_printf_format() implementation

musl does not provide parse_printf_format(). Let's introduce a fallback
method.

Co-authored-by: Yu Watanabe <watanabe.yu+github@gmail.com>
(cherry picked from commit 3bc03c67f0ea345e7fad49ed22d1e72bd833c1f9)

yuwata committed:
musl: introduce GNU specific version of strerror_r()

musl provides XSI compliant strerror_r(), and it is slightly different
from the one by glibc.
Let's introduce a tiny wrapper to convert XSI strerror_r() to GNU one.

The wrapper also patches musl's spurious catchall error message.

(cherry picked from commit ebb422dd2b3695de3abfb2dc4874bda432ca0c57)

yuwata committed:
musl: introduce dummy function for gnu_get_libc_version()

As the header gnu/libc-version.h and gnu_get_libc_version() function
are glibc specific, and musl does not provide them.

(cherry picked from commit 03d0fa4e4f824b725e3f250f70f4725c569c60a4)

yuwata committed:
musl: format-util: use %llu for formatting rlim_t

glibc uses uint32_t or uint64_t for rlim_t, while musl uses unsigned long long.

(cherry picked from commit a580dd4e53d218ceb7cb4cad46e7044dfa59bd48)

yuwata committed:
musl: meson: check existence of renameat2()

musl-1.2.5 does not provide renameat2(). Note, it is added by
https://github.com/kraj/musl/commit/05ce67fea99ca09cd4b6625cff7aec9cc222dd5a,
hence hopefully it will be provided by musl-1.2.6 or newer.

(cherry picked from commit 17e343b58b862306454b64cd66a26588420cffd3)

yuwata committed:
musl: meson: make musl not define wchar_t in their header

Otherwise, musl defines wchar_t as int, which conflicts with the
assumption by sd-boot, i.e. wchar_t is 2 bytes.

Fixes the following build error:
```
In file included from ../src/boot/efi-log.h:4,
                 from ../src/boot/linux_x86.c:13:
../src/boot/efi.h:19:24: error: conflicting types for 'wchar_t'; have 'short unsigned int'
   19 | typedef __WCHAR_TYPE__ wchar_t;
      |                        ^~~~~~~
In file included from /usr/include/stddef.h:19,
                 from ../src/boot/efi.h:9:
/usr/include/bits/alltypes.h:10:13: note: previous declaration of 'wchar_t' with type 'wchar_t' {aka 'int'}
   10 | typedef int wchar_t;
      |             ^~~~~~~
```

(cherry picked from commit 0736854da9dcfd06972b34aa7566e46e2997ca39)

yuwata committed:
musl: meson: explicitly set _LARGEFILE64_SOURCE

glibc sets it when _GNU_SOURCE is defined, however, musl does not.
Let's explicitly define it to make getdents64() and struct dirent64
available even when building with musl.

(cherry picked from commit fb33d20c072f2e7c3d500f00f8c04c99271f34c1)

yuwata committed:
musl: introduce dummy functions for mallinfo(), malloc_info(), and malloc_trim()

These functions are not provided by musl.

(cherry picked from commit bc610c70afcfe3cc489ecc18178f5d62d74e0d9b)

poettering commented on NEWS:
```diff
@@ -76,6 +76,10 @@ CHANGES WITH 258 in spe:
 
+        * systemd-sysusers now unconditionally reads and updates gshadow file
+          when necessary. The meson option 'gshadow' has been deprecated and
+          will be removed in a future release.
```
hmm, so we added this in #5920 as gshadow was optional back then apparently, and opensuse opted to disable it? was it actually? is it not anymore?

yuwata replied:
Hm, I did not know that. I will drop the change.

poettering commented on src/basic/dirent-util.h:
```diff
@@ -28,30 +28,12 @@ struct dirent* readdir_no_dot(DIR *dirp);
-/* Musl provides posix_getdents(). But glibc does not, and provides their own implementation as getdents64().
- * Let's introduce a simple wrapper. */
-#if !HAVE_POSIX_GETDENTS
-static inline ssize_t posix_getdents(int fd, void *buf, size_t nbyte, int flags) {
```
i think sticking to posix_getdents() makes a ton of sense, given it is just a wrapper around getdents64() but part of the posix spec, hence there's a chance glibc will eventuall support this too, in which case we can drop the whole thing. but reworking our code that consumes the function to something else seems unnecessary.

yuwata replied:
Currently, it is available only on musl git. Even the latest release of musl does not have it.
Also, by defining `_LARGEFILE64_SOURCE` musl also has `getdents64()` and `struct dirent64`. Hence, this is simply not necessary. Even if it is defined by posix.

poettering commented on src/libsystemd/sd-event/sd-event.c:
```diff
@@ -1874,13 +1872,19 @@ _public_ int sd_event_trim_memory(void) {
                 log_debug("Couldn't trim any memory.");
+#else
+        log_debug("malloc_trim() is not supported.");
+#endif
```
I actually would prefer if we'd add missing_malloc.h or so, which contains a static inline malloc_trim() that just returns zero or so.

poettering commented on src/shared/bus-util.c:
```diff
@@ -818,6 +818,7 @@ int bus_reply_pair_array(sd_bus_message *m, char **l) {
         return sd_bus_send(NULL, reply, NULL);
 }
 
+#if HAVE_MALLOC_INFO
```
actually would prefer if malloc_info() would just exist as a replacement call that returns EOPNOTSUPP if the libc is too dumb

poettering commented on src/basic/include/net/if.h:
```diff
@@ -6,5 +6,9 @@
 
+#ifndef __THROW
+#define __THROW
+#endif
```
urks, i wasn't aware we also carry copies of glibc headers. do we really want to do that? they define an .so api after all, not a syscall api. The failure modes if something is absent are quite different hence... syscall apis have a nice failure mode, they return ENOSYS. But .so apis do not really have that.

yuwata replied:
This is for musl.
With glibc, the macro is defined by glibc's features.h, and we use the macro as is in that case. There should be no problem.

poettering commented on src/basic/fileio.c:
```diff
@@ -165,13 +165,15 @@ int write_string_stream_full(
                 line = strjoina(line, "\n");
                 needs_nl = false;
         }
+#endif
```
this is a clear bug in musl, see https://mastodon.social/@cas@treehouse.systems/114654549930275910

kcxt replied:
the description for this commit is also entirely wrong afaict, the bug is absolutely not that musl writes 1 byte at a time

yuwata replied:
> the description for this commit is also entirely wrong afaict, the bug is absolutely not that musl writes 1 byte at a time

I am not sure. I just guess that by the commit message of the patch:
https://github.com/yoctoproject/poky/blob/5132c991e648d9ae8a6701d9da9e80bec65f0d25/meta/recipes-core/systemd/systemd/0009-Do-not-disable-buffering-when-writing-to-oom_score_a.patch

poettering commented on src/test/test-glob-util.c:
```diff
@@ -51,32 +57,44 @@ TEST(glob_exists) {
         const char *fn;
 
         _cleanup_globfree_ glob_t g = {
+#ifdef GLOB_ALTDIRFUNC
```
just dropping seems quite problematic. the filtering of "." and ".." kinda matters

yuwata replied:
fallback logic is implemented.

poettering commented on src/basic/parse-printf-format.h:
```diff
@@ -0,0 +1,55 @@
+#define	PA_FLAG_SHORT		(1 << 10)
+#define	PA_FLAG_PTR		(1 << 11)
+
+size_t parse_printf_format(const char *fmt, size_t n, int *types);
```
urks, this really should be maintained in a libmusl-compat library or so, we shouldn't maintain such complex glue code.

poettering commented on src/test/test-stat-util.c:
```diff
@@ -37,9 +37,15 @@ TEST(statx_definitions) {
         ASSERT_EQ(STATX_MNT_ID_UNIQUE,  0x00004000U);
+#ifdef STATX_SUBVOL
         ASSERT_EQ(STATX_SUBVOL,         0x00008000U);
+#endif
```
why doesn't linux/stat.h cover these? we carry our own copy of them after all

yuwata replied:
musl's sys/stat.h conflicts with linux/stat.h, hence we cannot include linux/stat.h here.
glibc's sys/stat.h internally includes linux/stat.h, and our tree includes the latest linux/stat.h, hence all constants are available.

poettering commented on src/test/test-cpu-set-util.c:
```diff
@@ -4,6 +4,10 @@
 
+#ifndef __GLIBC__
+typedef unsigned long __cpu_mask;
+#endif
```
hmm, this might not be correct. I think we should just skip the relevant test on non-glibc, that would be safer. 

it's better to not test this facet than to test the wrong thing

poettering commented on src/test/test-condition.c:
```diff
@@ -668,6 +671,7 @@ TEST(condition_test_version) {
         ASSERT_OK_ZERO(condition_test(condition, environ));
         condition_free(condition);
 
+#if HAVE_GNU_LIBC_VERSION_H
```
hmm, we should have a test here that verifies that checking the glibc version results in somewhat systematic behaviour if glibc is not used. probably the conditoin should always fail

poettering commented on src/basic/forward.h:
```diff
@@ -305,6 +307,7 @@ typedef struct VeritySettings VeritySettings;
 /* We duplicate various commonly used constants here so we can keep most static inline functions without
  * having to include the full header that provides these constants. */
 
+#undef AT_FDCWD
```
hmm, why undef? i'd prefer if we did #ifndef here

yuwata replied:
musl does
```c
#define AT_FDCWD (-100)
```
while we and glibc do
```c
#define AT_FDCWD -100
```
Hence, the definitions conflict each other.

yuwata committed:
musl: add several missing statx macros

glibc's sys/stat.h includes linux/stat.h, and we have copy of it from
the latest kernel, hence all new flags are always defined.
However, musl's sys/stat.h does not include linux/stat.h, and moreover,
they conflict with each other, hence we cannot include both header
simultaneously. Let's define missing macros to support musl.

(cherry picked from commit 69dd6d94200abf7010bc8c85cfe294e959014058)

yuwata committed:
musl: meson: gracefully disable gshadow, nss, and idn support

- musl does not support gshadow, and does not provide gshadow.h,
- musl does not support nss, and does not provide nss.h which is necessary
  for each nss modules,
- musl does not provide NI_IDN.

(cherry picked from commit 7dad0db2ee5a52d7ef7aca70f4d97ffd70212b12)

yuwata committed:
musl: provide several missing definitions for prctl()

(cherry picked from commit 6400e8dbd96bb45e88e8e6b4e035fe186a5ab743)

yuwata committed:
musl: replace netinet/if_ether.h with our own implementation

musl's netinet/if_ether.h conflicts with linux/if_ether.h.
The reimplementation is mostly equivalent with what glibc does.

(cherry picked from commit 8324ef421373c84b5034bf47b84fe42d84b1032f)

yuwata committed:
musl: avoid multiple evaluations in CPU_ISSET_S() macro

musl's CPU_ISSET_S() macro does not avoid multiple evaluations, and it
only accepts simple variable or constant.

Fixes the following error.
```
../src/shared/cpu-set-util.c: In function ‘cpu_set_to_mask_string’:
../src/shared/cpu-set-util.c:101:41: warning: operation on ‘i’ may be undefined [-Werror=sequence-point]
  101 |                         if (CPU_ISSET_S(--i, c->allocated, c->set))
      |                                         ^
```

(cherry picked from commit ebbc0ea7fd764c870bb8689111138b4f73854d42)

yuwata committed:
musl: meson: allow to choose libc implementation

This also introduces skeleton directories for storing musl specific code.

(cherry picked from commit a4c45279a76235f2cc9c298b6a30e3de668cb00b)

daandemeyer commented on src/include/override/sys/quota.h:
```diff
@@ -3,6 +3,10 @@
 
+/* glibc's sys/quota.h includes linux/quota.h since glibc-2.25 (4d728087ef8cc826b05bd21d0c74d4eca9b1a27d),
+ * but musl's one does not. Let's explicitly include it here. */
+#include <linux/quota.h>
```
I think we can upstream this patch by explicitly including linux/quota.h in every source file where it's required. Depending on transitive includes is not great, regardless of whether we're building with libc or musl so I'd be happy to accept a patch that adds a linux/quota.h include everywhere where needed.

yuwata replied:
Ugh... OK. Will do, even though I prefer the current approach, as I explained the reason at many times at other places.

daandemeyer replied:
I mean you can keep this musl specific patch as well ofc, I'm just saying that if you want to get rid of this musl specific patch, including linux/quota.h where needed is one way to do that.

yuwata committed:
musl: avoid conflict between fcntl.h and basic-forward.h

glibc defines AT_FDCWD as -100, but musl defines it as (-100).
In forward.h, we also define AT_FDCWD as -100, hence musl's fcntl.h
conflicts with forward.h. This is for avoiding the conflict.

(cherry picked from commit 43e4c91fc0ccf290a1b326038b5982308aee693b)

yuwata committed:
fd-util: split out fallback logic for close_all_fds()

No functional change. Just refactoring.

With this change, we can test each logic directly without seccomp
or hiding procfs.

(cherry picked from commit 93d037dbd09bc95126008fb19e9b129b8236d7a0)

yuwata committed:
fd-util: use heap only when nwea() does not work

Then, we can avoid malloc in most cases.

(cherry picked from commit f9f8a352646c206e72d5afd4f4d712456cf5aad4)

yuwata committed:
musl: build-path: fix reading DT_RUNPATH or DT_RPATH

musl records DT_STRTAB as offset, rather than address. So, need to add
obtained bias to read runpath or rpath.

(cherry picked from commit 4b774c740ba35d16c8a32d66d8bd5747c03d0d1a)

yuwata committed:
musl: meson: explicitly link with libintl when necessary

On some musl based distributions dgettext() may be provided by libintl.so.
Hence, we need to add dependency in that case.

(cherry picked from commit bd19ffd9cb618b15cbd74110aeca2abab745fe9e)

yuwata committed:
test-mount-util: several cleanups

- use safe_fork() with common flags,
- introduce macro for checking privilege.

(cherry picked from commit 2ff79a81a8dd5f3222d2be44f0d8c3b774b765d8)

yuwata committed:
test-efi-string: migrate to use ASSERT_OK() and friends

(cherry picked from commit 5809c451e8a6413af27fd73b5b533d612e9740bd)

yuwata committed:
test-parse-util: use newlocale() and freelocale() to set temporarl locale

This also drops unnecessary call of strtod(), as it is a test for
safe_atod().

(cherry picked from commit 6e219a09c2353ddbf52a66536db6c2b5593a903b)

yuwata committed:
test-seccomp-util: several cleanups

- use safe_fork() with FORK_WAIT
- introduce CHECK_SECCOMP() macro about common checks,
- ignore ENOSYS from sched_setscheduler().

(cherry picked from commit abd0eb20e430eb9e7f1e422eb7500748540194c7)

yuwata committed:
musl: hostname-util: introduce LINUX_HOST_NAME_MAX

glibc defines HOST_NAME_MAX as 64 and our code rely on that, but musl
defines the constant as 255. Let's provide our own definition for the
maximum length.

(cherry picked from commit 5bb9063505fe64788e696dc133899231506866c6)

yuwata committed:
musl: core: there is one less usable signal when built with musl

musl internally reserves one more signal, hence we can only use 29
signals.

(cherry picked from commit 3866923a6831a95a2aeca9b6f25694ec18440951)

yuwata committed:
test-sysusers: show diffs in stderr

Otherwise, the diffs are eaten by 'meson test' command unless --verbose
option is specified.

(cherry picked from commit aa5aac9e40b17a0571d0bf92d93e3b2ece98777a)

yuwata committed:
test-udev: pass test_env to make the test use systemd-detect-virt in build directory

Otherwise, the test fails if the running host does not have
systemd-detect-virt.

(cherry picked from commit 87a87b02b92d6cabbb86fe2c5829d3824d4c1265)

yuwata committed:
musl: time-util: skip tm.tm_wday check

musl does not set tm_wday when it is explicitly requested.
The check is not necessary at all, it is just for safety.
Let's skip it when built with musl.

(cherry picked from commit 8d5b306bb9901c7744bd79c7ceb38dcd93a48743)

yuwata committed:
musl: time-util: make parse_gmtoff() accept extended timezone offset format

musl v1.2.5 does not support %z specifier in strptime(). Since
https://github.com/kraj/musl/commit/fced99e93daeefb0192fd16304f978d4401d1d77
%z is supported, but it only supports strict RFC-822/ISO 8601 format,
that is, 4 digits with sign (e.g. +0900 or -1400), but does not support
extended format: 2 digits or colon separated 4 digits (e.g. +09 or -14:00).
Let's add fallback logic to make it support the extended timezone spec.

(cherry picked from commit 3ac4d68498dd378e2b3acd2bb86f4700263532d0)

yuwata committed:
musl: make strtoll() accept strings start with dot

glibc accepts strings start with '.' and returns 0, but musl refuses
them. Let's accept them, as our code assumes the function accept such
strings.

(cherry picked from commit 46ea7c3e3289fd6704f5fb4cfbb1421523ca1c7a)

yuwata committed:
time-util: introduce SAVE_TIMEZONE macro

(cherry picked from commit e3f561a6247579360099a25d03155f75ca428b9b)

yuwata committed:
time-util: introduce parse_gmtoff() helper function

(cherry picked from commit 23407c1806b037ce627a60d60df85a824b23967e)

yuwata committed:
time-util: fix issues in parse_timestamp() and optimize performance

Previously, an input string ends with short timezone spec e.g. WET,
was parsed by setting $TZ environment variable to the timezone.
But the timezone might be different from the original local timezone,
thus the result might not follow the timezone change in the original
local timezone.

This makes the check of the short timezone spec with tzname[] earlier,
then it is not necessary to load another timezone file for e.g. WET,
and provides expected time.

This also make it use SAVE_TIMEZONE macro and drop use of forking
process. This makes greatly improve performance when parsing string
that contains timezone different from the current local timezone.

Unfortunately, there is still one corner case that our test fails.
When tzdata is built with rearguard enabled, then at least
Africa/Windhoek timezone does not provide correct time, but time shifted
1 hour from the original.

(cherry picked from commit 514fa9d39ae9935ef1e014a3dd48dd5856007df2)

yuwata committed:
musl: test: several random fixlets for unit tests

(cherry picked from commit aec0e63a4f8afcfb434e036355b8e8c19dc07666)

yuwata committed:
musl: ci: add build test and unit tests

(cherry picked from commit f115763f191c45dc761aab071705385e78d9a3c5)

yuwata committed:
test: skip several test cases when running in chroot

When we are running in chroot, safe_fork() with FORK_MOUNTNS_SLAVE fails
with the following eror:
```
Failed to set mount propagation to MS_SLAVE for all mounts: Invalid argument
```
Let's skip the test cases when we are running in chroot.

(cherry picked from commit 6a616e62ad43fb52eacf083508253aeb9f57da5d)

yuwata committed:
musl: introduce dummy gshadow header file for userdb

Even 'gshadow' meson option is disabled, src/shared/userdb.c and
src/shared/user-record-nss.c include gshadow.h unconditionally.
Let's introduce dummy header to make them compiled gracefully.

(cherry picked from commit a2b7bcce27997aa77c05288ff6a560ac2f1f677a)

yuwata committed:
libc: drop unnecessary __THROW attribute

It is for C++, not necessary for us.

(cherry picked from commit 907ce9c3153e6296fb4fd4ded0d6ae62c7296b24)

yuwata committed:
tree-wide: drop redundant inclusion of linux/prctl.h

sys/prctl.h anyway includes linux/prctl.h and actually these .c files
includes sys/prctl.h. Hence, it is not necessary to explicitly include
linux/prctl.h.

(cherry picked from commit 9b55c4b85972df3d224a5a0852334e54324ca12c)

yuwata committed:
musl: time-util: introduce get_tzname() helper function

musl leaves the DST timezone name unset if there is no DST.
The helper function maps that back to no DST.

(cherry picked from commit 6431f2e072eef1c706b716d3a902fefcb1e63c40)

YHNdnzj committed:
core/exec-invoke: do not check array being non-NULL

Addresses
https://download.copr.fedorainfracloud.org/results/packit/systemd-systemd-39680/fedora-rawhide-i386/09787959-systemd/builder-live.log.gz

(cherry picked from commit 1d73c596fa890c696a8d91945a8b9adb05926991)

yuwata committed:
musl: locale-util: introduce musl specific locale enumerator

Both add_locales_from_archive() and add_locales_from_libdir() are glibc
specific, and the logic cannot be applied when built with musl.

(cherry picked from commit b2a2f670aec83e83a8798cf901950cb1ae63b54d)

yuwata committed:
musl: locale-util: explicitly check existence of locale file

musl's newlocale() always provides a locale object even the requested
locale does not exist. Let's explicitly check the existence of the
requested locale file.

(cherry picked from commit 361beb82a5ff6580c63f1e6fa4999ae6ff8772ed)

yuwata committed:
test-bus-error: use STRERROR() at several more places

(cherry picked from commit 6b656a5b40f3a65e55d169b5920f8e2101d4655e)

yuwata committed:
musl: test-time-util: skip parsing back when currently unused timezone is set

(cherry picked from commit f8a3ae1be6e19da6800b1b961228cdcb38b36774)

YHNdnzj committed:
time-util: do not carry musl-specific fallback logic on glibc systems

Follow-up for 3ac4d68498dd378e2b3acd2bb86f4700263532d0

We have no sensible way to detect why strptime() fails, hence
the fallback path as it is now would fire on glibc systems too,
pointlessly. Let's guard it behind ifdeffery.

(cherry picked from commit 6dd412bc50b14274a37f5e0d89efdd534518eeaf)

yuwata committed:
musl: stdio: check if stream is writable earlier in fputs() and friends

(cherry picked from commit 384e88a238390073ba69d8f9b9837c65bf01648d)

yuwata committed:
musl: test-bus-error: drop ._need_free flag checks

Its value depends on how strerror_r() implemented, and the
implementations of the function in glibc and musl are actually
different. Let's drop the checks.

(cherry picked from commit bb7d5e52a2a89d48a3e8194c824def9effdf7436)

keszybz committed:
musl: split out script to setup build

This makes it easier to set up a local build with musl:
$ tools/setup-musl-build.sh build-meson
$ ninja -C build-meson

(cherry picked from commit 31d3b96293a646d8016df2f0536a2ebb688f99f4)

yuwata committed:
musl: utmpx: add several missing definitions

(cherry picked from commit 3ae7d8fd8731c3412df878e8d1e8af9a8e644576)

keszybz committed:
musl: drop unused temporary pkgonfig directory

(cherry picked from commit 6edbf6a2c7a68a8a57ba900c85b694a82f2917cc)

yuwata committed:
[v258] musl: disable libiptc support in build test

Note, libiptc support is removed in v259.
